iothub/device/src/InternalClient.cs

```diff
@@ -1167,20 +1167,48 @@ public async Task UpdateReportedPropertiesAsync(TwinCollection reportedPropertie
             await InnerHandler.SendClientTwinPropertyPatchAsync(bodyStream, cancellationToken).ConfigureAwait(false);
         }
 
-        internal void OnDesiredStatePatchReceived(IDictionary<string, object> patch)
+        internal async void OnDesiredStatePatchReceived(IDictionary<string, object> patch)
         {
             if (Logging.IsEnabled)
                 Logging.Info(this, string.Join(",", patch), nameof(OnDesiredStatePatchReceived));
 
             if (_desiredPropertyUpdateCallback != null)
             {
                 TwinCollection twinUpdate = JsonConvert.DeserializeObject<TwinCollection>(JsonConvert.SerializeObject(patch));
-                _desiredPropertyUpdateCallback(twinUpdate, _twinPatchCallbackContext);
+                await _desiredPropertyUpdateCallback(twinUpdate, _twinPatchCallbackContext);
             }
             else if (_writableClientPropertyUpdateCallback != null)
             {
                 var writableClientPropertyCollection = new WritableClientPropertyCollection(patch, PayloadConvention);
-                _writableClientPropertyUpdateCallback(writableClientPropertyCollection);
+
+                var propertiesToBeReported = new ClientPropertyCollection();
```

Comment:
Why give it a default instance instead of `null`?

Reply:
No reason. This was minimal memory usage, so I didn't necessarily optimize. null should also work, sure.